render-test/runner.cpp

```diff
@@ -0,0 +1,408 @@
+#include <mbgl/map/camera.hpp>
+#include <mbgl/map/map_observer.hpp>
+#include <mbgl/style/conversion/filter.hpp>
+#include <mbgl/style/conversion/layer.hpp>
+#include <mbgl/style/conversion/light.hpp>
+#include <mbgl/style/conversion/source.hpp>
+#include <mbgl/style/image.hpp>
+#include <mbgl/style/layer.hpp>
+#include <mbgl/style/light.hpp>
+#include <mbgl/style/style.hpp>
+#include <mbgl/style/rapidjson_conversion.hpp>
+#include <mbgl/util/chrono.hpp>
+#include <mbgl/util/io.hpp>
+#include <mbgl/util/image.hpp>
+#include <mbgl/util/run_loop.hpp>
+#include <mbgl/util/string.hpp>
+#include <mbgl/util/timer.hpp>
+
+#include <mapbox/pixelmatch.hpp>
+
+#include "metadata.hpp"
+#include "parser.hpp"
+#include "runner.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <regex>
+
+bool TestRunner::checkImage(mbgl::PremultipliedImage&& actual, TestMetadata& metadata) {
+    const std::string base = metadata.path.remove_filename().string();
+    metadata.actual = mbgl::encodePNG(actual);
+
+    if (actual.size.isEmpty()) {
+        metadata.errorMessage = "Invalid size for actual image";
+        return false;
+    }
+
+#if !TEST_READ_ONLY
+    if (getenv("UPDATE")) {
+        mbgl::util::write_file(base + "/expected.png", mbgl::encodePNG(actual));
+        return true;
+    }
+
+    mbgl::util::write_file(base + "/actual.png", metadata.actual);
+#endif
+
+    mbgl::PremultipliedImage expected { actual.size };
+    mbgl::PremultipliedImage diff { actual.size };
+
+    double pixels = 0.0;
+
+    for (const auto& entry: readExpectedEntries(base)) {
+        mbgl::optional<std::string> maybeExpectedImage = mbgl::util::readFile(entry);
+        if (!maybeExpectedImage) {
+            metadata.errorMessage = "Failed to load expected image " + entry;
+            return false;
+        }
+
+        metadata.expected = *maybeExpectedImage;
+
+        expected = mbgl::decodeImage(*maybeExpectedImage);
+
+        pixels = // implicitly converting from uint64_t
+            mapbox::pixelmatch(actual.data.get(), expected.data.get(), expected.size.width,
+                               expected.size.height, diff.data.get(), 0.16); // GL JS uses 0.1285
```

Comment:
post merging comment:
> 0.16); // GL JS uses 0.1285
This is significant change. 0.1285 is set because we had some failures passing under the radar.
Let's not remove node render tests until the tolerance is the same - we could have node render tests running on one bot at least.